File: 0561-array-partition/0561-array-partition.java
class Solution {
    public int arrayPairSum(int[] nums) {
        int max = nums[0], min = nums[0];
        int n = nums.length;

        for (int i = 1; i < n; i++) {
            if (max < nums[i]) max = nums[i];
            if (min > nums[i]) min = nums[i];
        }

        int l = max - min + 1;
        int[] c = new int[l];

        for (int i = 0; i < n; i++) {
            c[nums[i] - min]++;
        }

        int sum = 0;
        boolean s = false;

        for (int k = 0; k < l; k++) {
            if (c[k] != 0) {
                int a = c[k];

                if (s) a--;

                sum += ((a + 1) / 2) * (k + min);

                s = a % 2 == 1;
            }
        }

        return sum;
    }
}

File: 0561-array-partition/0561-array-partition.py
class Solution:
    def arrayPairSum(self, nums: List[int]) -> int:
        max_val = nums[0]
        min_val = nums[0]
        n = len(nums)

        for i in range(1, n):
            if max_val < nums[i]:
                max_val = nums[i]

            if min_val > nums[i]:
                min_val = nums[i]

        l = max_val - min_val + 1
        c = [0] * l

        for i in range(n):
            c[nums[i] - min_val] += 1

        sum = 0
        s = False

        for k in range(l):
            if c[k] != 0:
                a = c[k]

                if s:
                    a -= 1

                sum += ((a + 1) // 2) * (k + min_val)

                s = a % 2 == 1

        return sum